src/Simulation/Core/IQirMachine.cs

```diff
@@ -0,0 +1,197 @@
+            /// Creates a double-precision floating point argument value.
+            /// </summary>
+            /// <param name="value">The value of the argument.</param>
+            public Double(double value) => this.Value = value;
+        }
+
+        /// <summary>
+        /// A Pauli operator argument value.
+        /// </summary>
+        public class Pauli : ArgumentValue
+        {
+            /// <summary>
+            /// The value of the argument.
+            /// </summary>
+            public Core.Pauli Value { get; }
+
+            /// <summary>
+            /// Creates a Pauli operator argument value.
+            /// </summary>
+            /// <param name="value">The value of the argument.</param>
+            public Pauli(Core.Pauli value) => this.Value = value;
+        }
+
+        /// <summary>
+        /// A range argument value.
+        /// </summary>
+        public class Range : ArgumentValue
+        {
+            /// <summary>
+            /// The value of the argument.
+            /// </summary>
+            public QRange Value { get; }
+
+            /// <summary>
+            /// Creates a range argument value.
+            /// </summary>
+            /// <param name="value">The value of the argument.</param>
+            public Range(QRange value) => this.Value = value;
+        }
+
+        /// <summary>
+        /// A result argument value.
+        /// </summary>
+        public class Result : ArgumentValue
+        {
+            /// <summary>
+            /// The value of the argument.
+            /// </summary>
+            public Core.Result Value { get; }
+
+            /// <summary>
+            /// Creates a result argument value.
+            /// </summary>
+            /// <param name="value">The value of the argument.</param>
+            public Result(Core.Result value) => this.Value = value;
+        }
+
+        /// <summary>
+        /// A string argument value.
+        /// </summary>
+        public class String : ArgumentValue
+        {
+            /// <summary>
+            /// The value of the argument.
+            /// </summary>
+            public string Value { get; }
+
+            /// <summary>
+            /// Creates a string argument value.
+            /// </summary>
+            /// <param name="value">The value of the argument.</param>
+            public String(string value) => this.Value = value;
+        }
+
+        /// <summary>
+        /// An array argument value.
+        /// </summary>
+        public class Array : ArgumentValue
+        {
+            /// <summary>
+            /// The values of the argument.
+            /// </summary>
+            public IQArray<ArgumentValue> Values { get; }
+
+            /// <summary>
+            /// Creates an array argument value.
+            /// </summary>
+            /// <param name="values">The values of the argument.</param>
+            public Array(IQArray<ArgumentValue> values) => this.Values = values;
```

Comment:
Should the condition that every value in the array is the same type be enforced here, or later?

Reply:
I think it would be good to do it here.

Reply:
I added a check.